src/terminal_box.rs

```diff
@@ -406,72 +406,69 @@ where
                                 renderer.fill_quad(underline2_quad, line_color);
                             }
 
-                            if metadata.flags.contains(Flags::DOTTED_UNDERLINE) {
-                                let bottom_offset = style_line_height * 2.0;
-
+                            let mut draw_repeated = |rects: &[(f32, Option<f32>)]| {
                                 let full_width = self.end_x - self.start_x;
-                                let mut accu_width = 0.0;
-                                let mut dot_width = 2.0f32.min(full_width - accu_width);
-
-                                while accu_width < full_width {
-                                    dot_width = dot_width.min(full_width - accu_width);
-                                    let pos_offset = mk_pos_offset!(accu_width, bottom_offset);
-                                    let underline_quad =
-                                        mk_quad!(pos_offset, style_line_height, dot_width);
-                                    renderer.fill_quad(underline_quad, line_color);
-                                    accu_width += 2.0 * dot_width;
+                                let pattern_len: f32 = rects.iter().map(|x| x.0).sum();
+                                let mut pos = 0.0;
+                                let mut index = {
+                                    let in_pattern = (self.start_x) % pattern_len;
+
+                                    let mut sum = 0.0;
+                                    let mut index = 0;
+                                    for (i, rect) in rects.iter().enumerate() {
+                                        sum += rect.0;
+                                        if in_pattern < sum {
+                                            let width = sum - in_pattern;
+                                            if let Some(height) = rect.1 {
+                                                let pos_offset = mk_pos_offset!(pos, height);
+                                                let underline_quad =
+                                                    mk_quad!(pos_offset, style_line_height, width);
+                                                renderer.fill_quad(underline_quad, line_color);
+                                            }
+                                            index = i + 1;
+                                            pos += width;
+                                            break;
+                                        }
+                                    }
+                                    index
+                                };
+                                while pos < full_width {
+                                    let (width, x) = rects[index % rects.len()];
```

Comment:
You forgot to limit the width here. Easy to spot with `./ansi-colors.sh` and dashed underlines ;)

``` rust
let (mut width, x) = rects[index % rects.len()];
width = width.min(full_width - pos);
```

Also `pos` is not very descriptive IMHO. Maybe going back `accu_width` or something similar would be preferred. Feel free to rename `full_width` too if you feel like it.